man/man1/yardstick.1
.\"    This file is part of the FElt finite element analysis package.
.\"    Copyright (C) 1993-2000 Jason I. Gobat and Darren C. Atkinson
.\"
.\"    This program is free software; you can redistribute it and/or modify
.\"    it under the terms of the GNU General Public License as published by
.\"    the Free Software Foundation; either version 2 of the License, or
.\"    (at your option) any later version.
.\"
.\"    This program is distributed in the hope that it will be useful,
.\"    but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"    GNU General Public License for more details.
.\"
.\"    You should have received a copy of the GNU General Public License
.\"    along with this program; if not, write to the Free Software
.\"    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.TH YARDSTICK 1fe "07/03/94" "Version 2.21" "Finite Element Package"
.SH NAME
yardstick \- unit conversion and problem scaling application
.SH SYNOPSIS
.B yardstick
[\-il \fIlength unit\fR]
[\-if \fIforce unit\fR]
[\-ol \fIlength unit\fR]
[\-of \fIforce unit\fR]
[\-ls \fIlength scale factor\fR]
[\-fs \fIforce scale factor\fR]
[\fIfilename\fR]
.SH DESCRIPTION
\fIyardstick\fR is a program to scale the numerical quantities in a FElt input
file; converting FElt files from one set of units to another
(e.g., Newtons and meters to pounds and inches) is a special case of this
operation and as such is given special treatment.  For unit conversion you
must specify input and output unit names for force and length quantities;
\fIyardstick\fR will figure out the appropriate scale factors. Mass units are
converted based on the length and force units that you specify.
Time units are always assumed to be seconds so no conversion options 
are provided.  For simple scaling you simply specify multiplicative
scale factors for both force quantities and length quantities.  Mass and time
quantities are treated as in the special case of unit conversion.
.SH OPTIONS
\fIyardstick\fR accepts the following options.  Note that for unit conversion
all four of \fI-il\fR, \fI-if\fR, \fI-ol\fR, \fI-of\fR must be specified and
that unit conversion and simple scaling (\fI-fs\fR and \fI-ls\fR) are exclusive
options.
.TP 5
.B \-help
Write a message describing valid command line options to standard error
and exit.
.TP
.B -il (--input-length) \fIlength units\fR
The symbolic name of the length units used in the input file.  Valid names
are m (meters), mm (millimeters), cm (centimeters), ft (feet), in (inches).
.TP
.B -if (--input-force) \fIforce units\fR
The symbolic name of the force units used in the input file.  Valid names
are N (Newtons), MN (mega-Newtons), GN (giga-Newtons), lbs (pounds),  
kips (kilopounds).
.TP
.B -ol (--output-length) \fIlength units\fR
The symbolic name of the length units for the output (i.e., what the input
units will be converted to).  Available types are the same as for input
length units.
.TP
.B -of (--output-force) \fIforce units\fR
The symbolic name of the force units for the output (i.e., what the input
units will be converted to).  Available types are the same as for input
length units.
.TP
.B -ls (--length-scale) \fIlength quantity scale factor\fR
The multiplicative factor by which length quantities within the input file
will be scaled.
.TP
.B -fs (--force-scale) \fIforce quantity scale factor\fR
The multiplicative factor by which force quantities within the input file
will be scaled.
.TP
.B \-nocpp
Do not use a preprocessor on the input file.
.TP
.BI \-cpp " filename"
Use \fIfilename\fR as a preprocessor on the input file.  The default
preprocessor is "/lib/cpp".  Any preprocssor which understands the -D, -U,
and -I options can be used as these options are passed to the preprocessor.
.SH AUTHOR
\fIyardstick\fR was developed by Jason I. Gobat (jgobat@mit.edu) and Darren
C. Atkinson (atkinson@ucsd.edu).
.SH SEE ALSO
felt(4fe).


CHANGELOG
** Changes to FElt from 3.02 to 3.05 (first version released on SourceForge)

The following changes were made to the FElt syntax:
   1. changed the sense of coordinates for axisymmetric elements to
      make them easier to use with mesh generation.  x is still radial
      but now y is vertical/axial.  Previously z was vertical.  

The following bugs/things were fixed/changed:
   1. some small bugs in axisymmetric, ctg, htk, and iso_2d elements
      that I haven't been really good about keeping tack of.  Try
      the new version and see if your answers look better -- if they
      look worse then let me know ...
   2. you can now use the -details switch to get the coordinates of
      the points where stresses are calculated in the text display of
      results.
   3. added radial and axial directions to the load dialog in velvet
      and to the file writing routines.  They'd gotten left out when
      axisymmetric elements were originally implemented.
   4. finally resolved the colormap issue in velvet so you can 
      use color stress and displacement plots on > 8-bit displays.
      Take a look at src/Velvet/XCC.README to see how you too can
      accomplish this if you're X11 colormap impaired like us.
   5. dropped Geompk in favor of Triangle by J.R. Shewchuk.  Several
      reasons for the change: Geompk was always a little finicky,
      mostly because of our interface probably, but it was sometimes
      difficult to coax a mesh out of Geompk.  The interface to
      Triangle is very simple, it is coded in C which makes compilation
      and configuration on other architectures simple, and it seems
      to be pretty easy to get a decent mesh out of it.  Accordingly,
      the main input parameters have changed to target and alpha.
      target is the approximate number of desired elements in the mesh.
      alpha is used to generate a maximum bound on the area of a generated
      element: Amax of an elt = alpha*Atotal/target.  Triangle appears to
      have a lot of other options for mesh refinement, angle constraints
      etc., but none of these are made use of at this point.  Most of the
      hooks for Geompk are still available if someone really wants to
      switch back.  The call to GenerateTriMesh did not change for 
      applications, but the applications did change to account for
      the loss of Geompk parameters (min, max, angspc, angtol, etc.)
      and the addition of target and alpha.
     
** Changes to FElt from 3.00 or 3.01 to 3.02 

The following changes were made to the FElt syntax:
   1. hooks for a gravity vector in the analysis section.  Doesn't
      do anything yet, but the syntax is there ...
   2. "static-substitution" and "static-incremental" are now valid 
      analysis types
   3. a "load cases" section and accompanying syntax for named
      load cases
   4. syntax for new analysis parameters, "tolerance", "iterations",
      "relaxation", "load-steps".
   5. input-node= and input-dof= parameters in the analysis 
      parameters to allow for parametric studies in which you
      get a table (or plot) of output that explores displacements
      as a function of force level at a single input point
   6. axial and radial are now valid directions for distributed loads

The following bugs/things were fixed/changed:
   1. the transform matrix for beam3d elements was incorrect if elements
      had nodes which were out of the x-y plane.
   2. several problems with the new Makefile scheme were corrected.
   3. velvet was probably getting the aspect ratios wrong in drawing 
      3d structures due to an error in the min/max checking.
   4. the spectrum components of force structures were not being
      initialized to either zero or NULL
   5. save's in velvet were actually doing dumps (save all's), minor, 
      but potentially annoying.
   6. there was still a problem with the lumped mass matrices for 
      htk elements.
   7. the new PrintGlobalMatrices routine took a file pointer but
      didn't actually use it; things were just going to stdout.
   8. internal error in burlap's interface wit the matrix lib
      was causing some problems with inversion and overwriting.
   9. there were problems with iso_2d elements (or any element with
      optional nodes) as the renumbering routines were just blindly
      looping through element -> definition -> numnodes rather than
      checking to make sure that all of the nodes were in fact valid.
  10. on a similar note, the checks in velvet for out-of-plane nodes
      were going over element -> definition -> numnodes rather than
      element -> definition -> shapenodes (which all have to be valid
      in the first place).
  11. there was some sort of problem with the pre-processor and
      corduroy (don't ask me how these things happen).  I think it's
      fixed now, but then again, I don't really know what was wrong.
  12. renamed AssembleStiffness and AssembleDynamic to ConstructStiffness
      and ConstructDynamic to reflect that they both create and assemble
      the matrices (this then parallels ConstructForceVector).  Renamed
      ConstructTransientForce to AssembleTransientForce to reflect 
      the fact it only fills the thing rather than creating it then
      filling it.
  13. the cpp command line wasn't being built properly so you couldn't
      actually use -I, -D and -U on the felt (and corduroy) command line
  14. memory initialization problem in Drawing.c
  15. 2d and 3d Athena widget specific defaults were moved into
      velvet.c and are now applied automatically at runtime as
      appropriate.
  16. the filename printing and scrolling in velvet's file dialog
      was improved.
  17. changed the analysis.nodes array into an array of actual node
      pointers rather than node numbers.  In some ways this adds
      a layer of complexity because whenever we input analysis nodes
      we have to resolve their numbers by doing tree lookups in the

src/Yardstick/yardstick.c
/*
    This file is part of the FElt finite element analysis package.
    Copyright (C) 1993-1997 Jason I. Gobat and Darren C. Atkinson

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

/****************************************************************************
 *
 * File:         yardstick.c
 *
 * Description:  Contains code for the driver application for the unit
 *               conversion package for FElt
 *
 *****************************************************************************/

# include <stdio.h>
# include <string.h>
# include "problem.h"
# include "error.h"
# include "allocate.h"
# include "definition.h"
# include "yardstick.h"
# include "units.h"

# define streq(a,b)	!strcmp(a,b)

double atof ( );

# ifndef DOS
static char *usage = "\
unit conversion: yardstick -if [force units] -il [length units]\n\
                           -of [force units] -ol [length units]\n\
                           [Cpp options] [filename]\n\
\n\
problem scaling: yardstick -fs [force scale factor] -ls [length scale factor]\n\
                           [Cpp options] [filename]\n\
\n\
       Available force units:    lbs, kips, N, MN, GN\n\
       Available length formats: in, ft, m, mm, cm\n\
\n\
       Cpp options:\n\
       -nocpp              do not use a preprocessor\n\
       -cpp filename       preprocessor to use\n\
       -Dname[=value]      define a macro\n\
       -Uname              undefine a macro\n\
       -Idirectory         specify include directory\n\
";
# else
static char *usage = "\
unit conversion: yardstick -if [force units] -il [length units]\n\
                           -of [force units] -ol [length units]\n\
                           [filename]\n\
\n\
problem scaling: yardstick -fs [force scale factor] -ls [length scale factor]\n\
                           [filename]\n\
\n\
       Available force units:    lbs, kips, N, MN, GN\n\
       Available length formats: in, ft, m, mm, cm\n\
\n\
";
# endif

static char *ilength = NULL;
static char *iforce  = NULL;

static char *olength = NULL;
static char *oforce  = NULL;

static double fscale = UnspecifiedValue;
static double lscale = UnspecifiedValue;

/************************************************************************
 * Function:	ParseOptions						*
 *									*
 * Description:	Parses the command line options.			*
 ************************************************************************/

static int ParseOptions (argc, argv)
    int  *argc;
    char *argv [ ];
{
    int   	i;
    int   	j;
    char 	*arg;

    j = 1;
    for (i = 1; i < *argc; i ++) {
	if (streq ((arg = argv [i]), "-help")) {
	    fputs (usage, stderr);
	    exit (0);
	} else if (streq (arg, "-il") || streq (arg, "--input-length")) {
	    if (++ i == *argc) 
		return 1;

           ilength = argv [i]; 
        } else if (streq (arg, "-if") || streq (arg, "--input-force")) {
            if (++ i == *argc) 
               return 1;

           iforce = argv [i];
        } else if (streq (arg, "-of") || streq (arg, "--output-force")) {
            if (++ i == *argc) 
               return 1;

           oforce = argv [i];
	} else if (streq (arg, "-ol") || streq (arg, "--output-length")) {
	    if (++ i == *argc) 
		return 1;

           olength = argv [i]; 
	} else if (streq (arg, "-fs") || streq (arg, "--force-scale")) {
	    if (++ i == *argc) 
		return 1;

           fscale = atof (argv [i]); 
	} else if (streq (arg, "-ls") || streq (arg, "--length-scale")) {
	    if (++ i == *argc) 
		return 1;

           lscale = atof (argv [i]); 
	} else
	    argv [j ++] = arg;
    }

    argv [*argc = j] = NULL;
    return 0;
}


/************************************************************************
 * Function:	 main							*
 *									*
 * Description:	 Main is the driver function for the yardstick 		*
 *		 application.						*
 ************************************************************************/

int main (argc, argv)
    int   argc;
    char *argv [ ];
{
    int		mode;
    unsigned	i;

	/*
	 * just in case there are cpp directives in the input
	 */

# ifndef DOS
    if (ParseCppOptions (&argc, argv)) {
        fputs (usage, stderr);
        exit (1);
    }
# endif
	/*
	 * get the yardstick specific options
	 */

    if (ParseOptions (&argc, argv)) {
	fputs (usage, stderr);
	exit (1);
    }

  	/*
	 * check the validity of all the possible option combinations
 	 * (and if this isn't a silly way to do this ...)
	 */

    mode = 0;
    if (oforce) mode ++;
    if (iforce) mode ++;
    if (olength) mode ++;
    if (ilength) mode ++; 

    if (fscale != UnspecifiedValue) mode += 10;
    if (lscale != UnspecifiedValue) mode += 10;

    if (mode != 4 && mode != 20) {
       if (mode < 4)
          Fatal ("must specify input and output force and length units.");

       if (mode == 10)
          Fatal ("must specify length and force scale factors.");

       Fatal ("scaling and unit conversion are exclusive options.");
    }

	/*
	 * figure out the scale factors based on the given units
	 * if we are doing unit conversion and not just scaling
	 */

    if (mode == 4) {
       fscale = 1.0;
       for (i = 0 ; i < sizeof (force_units) / sizeof (Unit) ; i++) {
          if (streq (force_units [i].name, iforce))
             fscale /= force_units [i].scale;
     
          if (streq (force_units [i].name, oforce))
             fscale *= force_units [i].scale;
       }

       lscale = 1.0;
       for (i = 0 ; i < sizeof (length_units) / sizeof (Unit) ; i++) {
          if (streq (length_units [i].name, ilength))
             lscale /= length_units [i].scale;
     
          if (streq (length_units [i].name, olength))
             lscale *= length_units [i].scale;
       }
    }

    add_all_definitions ( );

    	/*
	 * read the input file	
	 */ 

    if (ReadFeltFile (argc == 2 ? argv [1] : "-"))
       exit (1);
   
	/*
	 * scale all the numerical quantities in the input
	 */

    ScaleFeltFile (lscale, fscale);

	/*
	 * dump it back out to standard output
	 */

    DumpFeltFile ("-");

    exit (0);
}
